### DISCUSSION TITLE
Understanding type-narrowing behavior in 1.1.181

### DISCUSSION THREAD

**Discussion by rsokl** [microsoft/pyright#2480] [Q&A] [answered]
When upgrading to pyright 1.1.181, one of my functions started to fail that pyright's 'basic' checks, and I don't understand why.

A simplified example is:

```python
import inspect
from typing import Union, TypeVar

_T = TypeVar("_T")

def f(x: _T) -> Union[_T, int]:
    if callable(x) and inspect.isfunction(x):
        return 1
    return x  # <-- this line gets flagged
``` 

pyright-basic flags this with

```
error: Expression of type "object* | <callable subtype of object>" cannot be assigned to return type "_T@f | int"
    Type "object* | <callable subtype of object>" cannot be assigned to type "_T@f | int"
      Type "<callable subtype of object>" cannot be assigned to type "_T@f | int"
        Type "<callable subtype of object>" cannot be assigned to type "_T@f"
        "<callable subtype of object>" is incompatible with "int" (reportGeneralTypeIssues)
```

It isn't clear to me if this is a false-positive, or if I simply don't understand what this type-issue is. All of the following are  OK according to pyright-basic.

```python
def f(x: _T) -> Union[_T, int]:
    if callable(x):  # <- no is_function
        return 1
    return x
``` 

```python
def f(x: _T) -> Union[_T, int]:
    if inspect.isfunction(x):  # <- no `callable`
        return 1
    return x
``` 

```python
def f(x: _T) -> Union[_T, int]:
    if callable(x) and inspect.isfunction(x):
        return 1
    return cast(_T, x)   # <- add casting
``` 

This is not particularly pressing, but I would like to have a better understanding of what is going on here. Thank you!

**Comment by erictraut** [comment, marked as answer]
This looks to be a regression related to the fix I added for https://github.com/microsoft/pyright/issues/2472. Sorry about that. I'll work on a fix for this one and augment our unit tests to cover this case.

**Comment by rsokl** [comment]
No problem!

Please feel free to delete this post from the discussion board, since it isn't really relevant for future readers / it should have been an Issue.
